extensions/amp-user-notification/0.1/amp-user-notification.js

```diff
@@ -0,0 +1,375 @@
+  getAsyncCid_() {
+    this.dialogPromise_ = new Promise((resolve, reject) => {
+      this.dialogResolve_ = resolve;
+    });
+    return cidFor(this.win_).then(cid => {
+      // `amp-user-notification` is our cid scope, while we give it a resolved
+      // promise for the 2nd argument so that the 3rd argument (the
+      // persistentConsent) is the one used to resolve getting
+      // the external CID.
+      // The dialogPromise_ is never rejected,
+      // the user only really has 1 option to accept/dismiss (to resolve)
+      // the notification or have the nagging notification sitting there
+      // (to never resolve).
+      return cid.get('amp-user-notification',
+          Promise.resolve(), this.dialogPromise_);
+    });
+  }
+
+  /** @override */
+  shouldShow() {
+    return this.getAsyncCid_()
+        .then(this.postShowEndpoint_.bind(this))
+        .then(this.onPostShowEndpointSuccess_.bind(this));
+  }
+
+  /** @override */
+  show() {
+    this.element.style.display = '';
+    this.element.classList.add('amp-active');
+    return this.dialogPromise_;
+  }
+
+  /**
+   * Hides the current user notification and invokes the `dialogResolve_`
+   * method. Removes the `.amp-active` class from the element.
+   */
+  dismiss() {
+    this.element.classList.remove('amp-active');
+    this.element.classList.add('amp-hidden');
+    this.dialogResolve_();
+    this.postDismissEnpoint_();
+  }
+}
+
+
+/**
+ * UserNotificationManager handles `amp-user-notification`
+ * queuing and registration, as well as exposing the components
+ * dismiss promise.
+ */
+export class UserNotificationManager {
+
+  constructor(window) {
+    this.win_ = window;
+
+    /** @private @const {!Object<!UserNotificationDeferDef>} */
+    this.deferRegistry_ = Object.create(null);
+
+    /** @private @const {!Viewer} */
+    this.viewer_ = viewerFor(this.win_);
+
+    /** @private @const {?function(*)} */
+    this.documentReadyResolve_ = null;
+
+    let documentReadyPromise_ = new Promise((resolve) => {
+      this.documentReadyResolve_ = resolve;
+    });
+
+    // TODO(erwinm, https://github.com/ampproject/amphtml/issues/1226)
+    // Add `whenDocumentReady` function in document-state
+    // that returns a promise.
+    onDocumentReady(this.win_.document, this.documentReadyResolve_);
+
+    /** @private {!Promise} */
+    this.managerReadyPromise_ = this.viewer_.whenVisible().then(() => {
+      // we block on both viewer visible and when dom ready.
+      return documentReadyPromise_;
+    });
+
+    this.nextInQueue_ = this.managerReadyPromise_;
+  }
+
+  /**
+   * Retrieve a promise associated to an `amp-user-notification` component
+   * that is resolved when user agrees to the terms.
+   * @param {string} id
+   * @return {!Promise<string>}
+   */
+  get(id) {
+    this.managerReadyPromise_.then(() => {
+      if (this.win_.document.getElementById(id) == null) {
+        console./*OK*/warn(`Did not find amp-user-notification element ${id}.`);
```

Comment:
@dvoytenko heres the latest iteration of the check.

i think doing the check in the register function makes the id (and element) guaranteed to be there since that is coming from the custom-element. the check needs to definitely happen here on the `get` as it is the externally exposed api to other services.